File: backend/services/stigmer-server/pkg/domain/mcpserver/oauth/dcr.go
package oauth

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// DCRRequest is the Dynamic Client Registration request per RFC 7591.
type DCRRequest struct {
	RedirectURIs            []string `json:"redirect_uris"`
	ClientName              string   `json:"client_name"`
	GrantTypes              []string `json:"grant_types"`
	ResponseTypes           []string `json:"response_types"`
	TokenEndpointAuthMethod string   `json:"token_endpoint_auth_method"`
}

// DCRResponse is the Dynamic Client Registration response per RFC 7591.
type DCRResponse struct {
	ClientID                string `json:"client_id"`
	ClientSecret            string `json:"client_secret,omitempty"`
	ClientName              string `json:"client_name,omitempty"`
	TokenEndpointAuthMethod string `json:"token_endpoint_auth_method,omitempty"`
}

var dcrHTTPClient = &http.Client{Timeout: 15 * time.Second}

// RegisterClient performs OAuth Dynamic Client Registration (RFC 7591)
// at the given registration endpoint.
//
// Registers a public client (token_endpoint_auth_method: "none") since
// MCP OAuth uses PKCE instead of client secrets. The returned client_id
// is stored in the user's OAuthGrant for subsequent token operations.
func RegisterClient(ctx context.Context, registrationEndpoint, redirectURI, clientName string) (*DCRResponse, error) {
	reqBody := DCRRequest{
		RedirectURIs:            []string{redirectURI},
		ClientName:              clientName,
		GrantTypes:              []string{"authorization_code"},
		ResponseTypes:           []string{"code"},
		TokenEndpointAuthMethod: "none",
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal DCR request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, registrationEndpoint, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create DCR request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	resp, err := dcrHTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("DCR request to %s failed: %w", registrationEndpoint, err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read DCR response: %w", err)
	}

	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf(
			"DCR at %s returned HTTP %d: %s",
			registrationEndpoint, resp.StatusCode, truncateBody(respBody),
		)
	}

	var dcrResp DCRResponse
	if err := json.Unmarshal(respBody, &dcrResp); err != nil {
		return nil, fmt.Errorf("failed to parse DCR response: %w", err)
	}

	if dcrResp.ClientID == "" {
		return nil, fmt.Errorf("DCR response from %s is missing client_id", registrationEndpoint)
	}

	return &dcrResp, nil
}

func truncateBody(body []byte) string {
	const maxLen = 256
	if len(body) <= maxLen {
		return string(body)
	}
	return string(body[:maxLen]) + "..."
}

File: backend/services/stigmer-server/src/domain/mcpserver/oauth/dcr.ts
/**
 * RFC 7591 Dynamic Client Registration — ports
 * pkg/domain/mcpserver/oauth/dcr.go. Registers a PUBLIC client
 * (token_endpoint_auth_method "none"): MCP OAuth uses PKCE instead of a
 * client secret, and an empty secret is what keeps the pending state's
 * "never ciphertext-of-empty" rule meaningful (oss#394).
 * Proven by mcpserver-oauth.conformance.test.ts (CONFORMANCE_TARGET=local).
 */
import { truncateBody } from "./truncate-body.js";

/** Dynamic Client Registration response per RFC 7591 (Go DCRResponse). */
export interface DcrResponse {
  clientId: string;
  clientSecret: string;
  clientName: string;
  tokenEndpointAuthMethod: string;
}

/** Go's dcrHTTPClient 15s timeout — a named constant per guidelines. */
export const DCR_REQUEST_TIMEOUT_MS = 15_000;

/**
 * Performs OAuth Dynamic Client Registration at the given registration
 * endpoint (Go RegisterClient). The returned client_id is stored in the
 * user's OAuthGrant for subsequent token operations. Accepts 200 or 201
 * (RFC says 201; real providers answer both).
 */
export async function registerClient(
  registrationEndpoint: string,
  redirectUri: string,
  clientName: string,
  fetchImpl: typeof fetch = fetch,
): Promise<DcrResponse> {
  const requestBody = JSON.stringify({
    redirect_uris: [redirectUri],
    client_name: clientName,
    grant_types: ["authorization_code"],
    response_types: ["code"],
    token_endpoint_auth_method: "none",
  });

  let response: Response;
  try {
    response = await fetchImpl(registrationEndpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: requestBody,
      signal: AbortSignal.timeout(DCR_REQUEST_TIMEOUT_MS),
    });
  } catch (error) {
    throw new Error(
      `DCR request to ${registrationEndpoint} failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  const body = await response.text();
  if (response.status !== 201 && response.status !== 200) {
    throw new Error(
      `DCR at ${registrationEndpoint} returned HTTP ${response.status}: ${truncateBody(body)}`,
    );
  }

  let raw: {
    client_id?: string;
    client_secret?: string;
    client_name?: string;
    token_endpoint_auth_method?: string;
  };
  try {
    raw = JSON.parse(body) as typeof raw;
  } catch (error) {
    throw new Error(
      `failed to parse DCR response: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  if ((raw.client_id ?? "") === "") {
    throw new Error(
      `DCR response from ${registrationEndpoint} is missing client_id`,
    );
  }

  return {
    clientId: raw.client_id as string,
    clientSecret: raw.client_secret ?? "",
    clientName: raw.client_name ?? "",
    tokenEndpointAuthMethod: raw.token_endpoint_auth_method ?? "",
  };
}
